e3nn/spherical_tensor.py

```diff
@@ -90,3 +96,283 @@ def spherical_harmonics_coeff_to_sphere(coeff, alpha, beta):
+        signal = adjusted_projection(vectors, L_max, sum_points=sum_points, radius=radius)
+        return cls(signal, Rs)
+
+    @classmethod
+    def from_geometry_with_radial(cls, vectors, radial_model, L_max, sum_points=True):
+        r = vectors.norm(2, -1)
+        radial_functions = radial_model(r)
+        _N, R = radial_functions.shape
+        Rs = [(R, L) for L in range(L_max + 1)]
+        mul_map = rs.map_mul_to_Rs(Rs)
+        radial_functions = torch.einsum('nr,dr->nd',
+                                        radial_functions.repeat(1, L_max + 1),
+                                        mul_map)  # [N, signal]
+
+        Ys = projection(vectors, L_max, sum_points=False, radius=False)  # [channels, N]
+        irrep_map = rs.map_irrep_to_Rs(Rs)
+        Ys = torch.einsum('nc,dc->nd', Ys, irrep_map)  # [N, signal]
+
+        signal = Ys * radial_functions  # [N, signal]
+
+        if sum_points:
+            signal = signal.sum(0)
+        new_cls = cls(signal, Rs)
+        new_cls.radial_model = radial_model
+        return new_cls
+
+    def sph_norm(self):
+        Rs = self.Rs
+        signal = self.signal
+        n_mul = sum([mul for mul, L in Rs])
+        # Keep shape after Rs the same
+        norms = torch.zeros(n_mul, *signal.shape[1:])
+        sig_index = 0
+        norm_index = 0
+        for mul, L in Rs:
+            for _ in range(mul):
+                norms[norm_index] = signal[sig_index: sig_index + (2 * L + 1)].norm(2, 0)
+                norm_index += 1
+                sig_index += 2 * L + 1
+        return norms
+
+    def signal_on_sphere(self, which_mul=None, n=100, radius=True):
+        n_mul = sum([mul for mul, L in self.Rs])
+        if which_mul:
+            if len(which_mul) != n_mul:
+                raise ValueError("which_mul and number of multiplicities is not equal.")
+        else:
+            which_mul = [1 for i in range(n_mul)]
+
+        # Need to handle if signal is featurized
+        x, y, z = (None, None, None)
+        Ys = []
+        for mul, L in self.Rs:
+            # Using cache-able function
+            x, y, z, Y = spherical_harmonics_on_grid(L, n)
+            Ys += [Y] * mul
+
+        f = self.signal.unsqueeze(1).unsqueeze(2) * torch.cat(Ys, dim=0)
+        f = f.sum(0)
+        return x, y, z, f
+
+    def plot(self, which_mul=None, n=100, radius=True, center=None, relu=True):
+        """
+        surface = self.plot()
+        fig = go.Figure(data=[surface])
+        fig.show()
+        """
+        import plotly.graph_objs as go
+
+        x, y, z, f = self.signal_on_sphere(which_mul, n, radius)
+        f = f.relu() if relu else f
+
+        if radius:
+            r = f.abs()
+            x = x * r
+            y = y * r
+            z = z * r
+
+        if center is not None:
+            x = x + center[0]
+            y = y + center[1]
+            z = z + center[2]
+
+        return go.Surface(x=x.numpy(), y=y.numpy(), z=z.numpy(), surfacecolor=f.numpy())
```

Comment:
Should we instead output `torch.stack([x, y, z], f)` similar to `plot_with_radial` (rather than a plotly plot)? Alternatively, we can have a switch for whether the raw data or a plotly plot is returned.

Reply:
From our conversation IRL we should just return [x, y, z], f

Reply:
where [x, y, z] = r

Reply:
I dont like swiches